Feature/refactor config

martonborzak opened the pull request:
Added 2 methods to extend the YIO API functionalities. These methods will help the web configurator to create a drag and drop management interface for entities.

`getAllAvailableEntities` returns a list of available entities. These entities are not controlled by the remote, but available from the integration.

`addAvailableEntity` adds an entity to the list of available entities. It checks if the entity type is supported. If so, it returns true, otherwise false. This is implemented in `integration` class and can be used by all integration implementations.

martonborzak committed:
refactor: Removed unused methods

martonborzak committed:
refactor: Changed getAllAvailableEntities return type

martonborzak committed:
refactor: Change type of getAllAvailableEntities implementation

martonborzak committed:
refactor: Change getAvailableEntities to QVariantList

martonborzak committed:
refactor: Added supported_entities virtual method to entitiesinterface

zehnm commented:
Please provide a description what this pull request contains and the purpose of the changes.
E.g. what is the reason to change `getAllAvailableEntities()`to now return a `QVariantList` instead of `QStringList`?
This is a breaking change which affects all integration plugins, but there are no pull requests yet.

zehnm commented on src/yio-interface/entities/entitiesinterface.h:
```diff
@@ -48,6 +48,9 @@ class EntitiesInterface {
     virtual EntityInterface* getEntityInterface(const QString& entity_id) = 0;
+
+    // get a list of supported entities
+    virtual QStringList supported_entities() = 0;
```
Use lowerCamelCase for naming: `supportedEntities`

zehnm commented on src/yio-interface/integrationinterface.h:
```diff
@@ -57,7 +57,7 @@ class IntegrationInterface {
      * @brief getAllAvailableEntities Can be implemented by integration
      */
-    virtual QStringList getAllAvailableEntities() = 0;
+    virtual QVariantList getAllAvailableEntities() = 0;
```
Why is this change required?
What are the actual types that can be returned? E.g. does a list of integers make sense?

zehnm commented on src/yio-plugin/integration.h:
```diff
@@ -100,5 +100,9 @@ class Integration : public QObject, public IntegrationInterface {
     QLoggingCategory&       m_logCategory;
-    QStringList             m_allAvailableEntities;
+    QVariantList            m_allAvailableEntities;
+    /// containts QVariantMap objects with the following keys
+    /// "entity_id": entity id
+    /// "type": type of entity, like light
+    /// "integration_id": id of the integration controlling the entity
```
I don't understand those comments. Are they related to the content of QVariantList?
Then the description must be added to the getter method in IntegrationInterface to describe the API.

martonborzak committed:
feat: added addAvailableEntity method

addAvailableEntity will add the entity to the available entities list if the entity type is supported

martonborzak committed:
refactor: Renamed supported_entities to supportedEntities

martonborzak committed:
fix: check if the new available entity is in the list

martonborzak committed:
feat: Added `isSupportedEntityType` to entitiesinterface

martonborzak committed:
refactor: use new `isSupportedEntityType` instead of stringlist

zehnm commented on src/yio-interface/integrationinterface.h:
```diff
@@ -57,7 +57,18 @@ class IntegrationInterface {
+    virtual QVariantList getAllAvailableEntities() = 0;
+
+    /**
+     * @brief sendCommand Must be implemented as Q_INVOKABLE
```
sendCommand?

zehnm replied:
Please fix documentation, looks like copy pasted.

zehnm commented on src/yio-interface/integrationinterface.h:
```diff
@@ -57,7 +57,18 @@ class IntegrationInterface {
+     * @param friendly_name
+     * @param supported_features
+     */
+    virtual bool addAvailableEntity(const QString& entity_id, const QString& type, const QString& integration,
+                                    const QString& friendly_name, const QStringList& supported_features) = 0;
```
please use lowerCamelCase for parameter naming

zehnm commented on src/yio-plugin/integration.cpp:
```diff
@@ -69,6 +69,30 @@ Integration::Integration(Plugin* plugin)
 
 Integration::~Integration() {}
 
+bool Integration::addAvailableEntity(const QString& entity_id, const QString& type, const QString& integration,
+                                     const QString& friendly_name, const QStringList& supported_features) {
```
lowerCamelCase for parameters as in header file

zehnm commented on src/yio-plugin/integration.h:
```diff
@@ -59,11 +59,13 @@ class Integration : public QObject, public IntegrationInterface {
+    void             enterStandby() {}  // Can be overriden by integration
+    void             leaveStandby() {}  // Can be overriden by integration
+    QVariantList     getAllAvailableEntities() { return m_allAvailableEntities; }
+    bool             addAvailableEntity(const QString& entity_id, const QString& type, const QString& integration,
+                                        const QString& friendly_name, const QStringList& supported_features);
```
lowerCamelCase naming

martonborzak committed:
refactor: Rename variables & add description to header file

martonborzak committed:
refactor: change CamelCase

martonborzak committed:
refactor: added missing overrides

martonborzak committed:
refactor: variable renaming

martonborzak committed:
refactor: Renaming in threadadapter

zehnm commented on src/yio-plugin/integration_threadadapter.h:
```diff
@@ -51,16 +51,16 @@ class IntegrationThreadAdapter : public Integration {
-    QStringList      getAllAvailableEntities() override;
-    Q_INVOKABLE void sendCommand(const QString& type, const QString& entity_id, int command,
+    QVariantList     getAllAvailableEntities() override;
+    Q_INVOKABLE void sendCommand(const QString& type, const QString& entityId, int command,
```
I don't think we need a Q_INVOKABLE here. It's an internal helper class.
In my opinion all Q_INVOKABLE should be slots including enterStandby and leaveStandby.

zehnm commented on src/yio-interface/integrationinterface.h:
```diff
@@ -56,23 +56,36 @@ class IntegrationInterface {
+                                    const QString& friendlyName, const QStringList& supportedFeatures) = 0;
+
+    /**
+     * @brief sendCommand Must be implemented as Q_INVOKABLE
```
Why does it need to be a Q_INVOKABLE?
Is there any QML dependency to the integration interface?

zehnm commented on src/yio-plugin/integration.h:
```diff
@@ -55,31 +55,34 @@ class Integration : public QObject, public IntegrationInterface {
-    QStringList      getAllAvailableEntities() { return m_allAvailableEntities; }  // Can be overriden by integration
-    Q_INVOKABLE void sendCommand(const QString& type, const QString& entity_id, int command, const QVariant& param) = 0;
+    Q_INVOKABLE void connect() override    = 0;  // Must be implemented by integration
+    Q_INVOKABLE void disconnect() override = 0;  // Must be implemented by integration
```
I don't see the reason for Q_INVOKABLE. Everything should be handled by C++ now afaik. Therefore defining them as slots would be better, or I'm still missing something.
Or do we really allow people to write integration plugins in QML?

martonborzak replied:
connect and disconnect is `Q_INVOKABLE`, so you can manually disconnect the integration from the UI. If you go to `Settings/Integrations` there is option to connect/disconnect.

zehnm reviewed (commented):
We need to go over the interfaces again, especially Q_INVOKABLE methods.
It would be nice to have a rather stable v1 version and not to break it again with the next release.

martonborzak committed:
refactor: removed Q_INVOKABLEs

martonborzak committed:
fix: cpplint open issue

zehnm reviewed (approved):
As discussed we continue cleaning up the interfaces and create a minor release now to continue with the web configurator.
